### ISSUE TITLE
Recentre atoms based on centre-of-mass of topology

### ISSUE THREAD

**Issue by debeshmandal** [readdy/readdy#216]
Is it possible to recentre the whole system so that all of the particles a translated by a distance vector `[x, y, z]` so that a single topology's centre of mass is at `[0, 0, 0]` ?

**Comment by clonker** [comment]
You can use checkpointing and then update a single checkpoint to fulfill that requirement - I assume this is a post processing step you are after? Or do you want to do this within a simulation?

**Comment by debeshmandal** [comment]
Thanks for quick response - for more detail

I want to create a ConvexHull (using `scipy`) around a topology which is does post-processing - however there is a problem when it crosses a periodic boundary (it is not possible to know if this happens before running the simulation, so it makes creating checkpoints at the correct interval difficult)

When using the simulation package lammps, there is a command [`fix recenter`](https://docs.lammps.org/fix_recenter.html) that uses the particle positions and periodic image to achieve the recentering during the simulation.

I have heard of a post-processing technique where molecules are rebuilt using their positions and topological information (and not using periodic image information) - but this for me is much slower to do in post-processing so if it is possible to do this within the simulation then that would be preferred.

**Comment by clonker** [comment]
If it is just post processing that you need you can figure out whether minimum image convention was applied between two timesteps by comparing the positions of particles: If the delta is larger than half the box size then the particle was wrapped, in which case you can add the delta back to the particle's position. 
If you want to do it during simulation you could implement your own integrator which does that and link it against readdy. 
Yet another possibility that is not quite as invasive as writing your own integrator would be having a structural topology reaction with a very high reaction rate so that it gets evaluated basically every timestep. In it you can do something like this (pseudocode):

```python
def structural_reaction(topology):
    com = np.array([0, 0, 0])
    for vertex in topology.vertices:
        com += vertex.position
    com /= len(topology.vertices)
    
    recipe = readdy.StructuralReactionRecipe(topology)
    for vertex in topology.vertices:
        recipe.change_particle_position(v, vertex.position - com)
    return recipe

system.topologies.add_structural_reaction(...., structural_reaction, reaction_rate)
```

**Comment by debeshmandal** [comment]
I guess the first method would be better since I have non-topological particles as well that must be translated as well. I will post a successful solution here if I can manage to solve my problem.

Out of curiosity .. How would one add to the existing integrator to include an option for recentering? e.g. I would assume that in [`/include/readdy/kernel/singlecpu/actions/SCPUEulerBDIntegrator.h`](https://github.com/readdy/readdy/blob/master/include/readdy/kernel/singlecpu/actions/SCPUEulerBDIntegrator.h) something would have to added as well as a new `Action` ?

**Comment by clonker** [comment]
Yes in case it is restricted to post-processing I would say the first method is the way to go. You can check whether you are doing it the wrap-detection correctly by computing the mean-squared-displacement of particles over time and comparing the slope to the diffusion constant (you can look up an example of this in the readdy2 publication).

To add a new integrator you'd have to implement [readdy::model::actions::UserDefinedAction](https://github.com/readdy/readdy/blob/master/include/readdy/model/actions/UserDefinedAction.h), export it to python and set it as integrator in the simulation object (like `simulation.integrator = MyCustomIntegrator()`). Due to namespace mangling etc it is beneficial if you compile readdy directly together with the custom integrator.
If that is something you'd like to try out I can send you a small toy project.

**Comment by debeshmandal** [comment]
I would be interested in taking a look - however I am not that sure if it will be  possible for me to implement - I am hoping that it wouldn't be too tricky to measure the COM at timestep=0 and then do the translation every timestep - but it would be cool to see a toy system

**Comment by clonker** [comment]
https://github.com/clonker/readdy_com_integrator this here should do as a starting point, you should be able to `python setup.py install` and then
```python
from readdy_com import COMIntegrator

...
timestep = 1e-4
simulation.integrator = COMIntegrator(timestep)
```

**Comment by debeshmandal** [comment]
Thanks for putting this together - it makes sense to me how it works - I will let you know if it works for me
